### PR TITLE
[Aks] Add addon support

### PR DESCRIPTION
**Pull request by wyunchi-ms**
## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Aks/Aks/Utils/AddonUtils.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+namespace Microsoft.Azure.Commands.Aks.Utils
+{
+    static class AddonUtils
+    {
+        public static string TrimWorkspaceResourceId(string workspaceResourceId)
+        {
+            return string.Format("/{0}", workspaceResourceId.Trim().Trim('/'));
+        }
+
+        public static IDictionary<string, ManagedClusterAddonProfile> EnableAddonsProfile(IDictionary<string, ManagedClusterAddonProfile> addonProfiles, string[] AddonName, string workspaceResourceId, string subnetName)
```
Comment (erich-wang) [inline src/Aks/Aks/Utils/AddonUtils.cs:32]: please use lower case for the parameter name `AddonName`

File: src/Aks/Aks/Commands/UpdateAddonsBase.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(
+            Position = 1,
+            Mandatory = true,
+            ParameterSetName = DefaultParamSet,
+            HelpMessage = "Kubernetes managed cluster Name.")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
```
Comment (erich-wang) [inline src/Aks/Aks/Commands/UpdateAddonsBase.cs:51]: Suggest to use `ClusterName` to be consistent with New-AzAksNodePool

File: src/Aks/Aks/Commands/UpdateAddonsBase.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+        [Parameter(Mandatory = true,
+            ParameterSetName = InputObjectParameterSet,
+            ValueFromPipeline = true,
+            HelpMessage = "A PSKubernetesCluster object, normally passed through the pipeline.")]
+        [ValidateNotNullOrEmpty]
+        public PSKubernetesCluster InputObject { get; set; }
```
Comment (erich-wang) [inline src/Aks/Aks/Commands/UpdateAddonsBase.cs:58]: Suggest to use ClusterObject as it is top level object

File: src/Aks/Aks/Commands/UpdateAddonsBase.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Id of a managed Kubernetes cluster.")]
+        [ValidateNotNullOrEmpty]
+        [Alias("ResourceId")]
+        public string Id { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Addon names to be enabled when cluster is created.")]
+        [ValidateNotNullOrEmpty()]
+        [PSArgumentCompleter("HttpApplicationRouting", "Monitoring", "VirtualNode", "AzurePolicy", "KubeDashboard")]
+        public string[] AddonName { get; set; }
```
Comment (erich-wang) [inline src/Aks/Aks/Commands/UpdateAddonsBase.cs]: Please use `Name`

File: src/Aks/Aks/Commands/UpdateAddonsBase.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                        Name = resource.ResourceName;
+                        break;
+                    }
+            }
+
+            var msg = $"{Name} in {ResourceGroupName}";
+
+            if (ShouldProcess(msg, Resources.UpdateOrCreateAManagedKubernetesCluster))
+            {
+                var queriedCluster = GetManagedClusterWithResourceGroupNameAndName();
```
Comment (erich-wang) [inline src/Aks/Aks/Commands/UpdateAddonsBase.cs:101]: How about not querying cluster object if using InputObjectParameterSet?

File: src/Aks/Aks/Commands/UpdateAddonsBase.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        {
+            try
+            {
+                var cluster = Client.ManagedClusters.Get(ResourceGroupName, Name);
+                WriteVerbose(string.Format(Resources.ClusterExists, cluster.Id));
+                return cluster;
+            }
+            catch (CloudException)
+            {
+                WriteVerbose(Resources.ClusterDoesNotExist);
```
Comment (erich-wang) [inline src/Aks/Aks/Commands/UpdateAddonsBase.cs:130]: WriteError?

File: src/Aks/Aks/Commands/NewKubeBase.cs [hunk 7]
```diff
@@ -61,6 +63,23 @@ public abstract class NewKubeBase : CreateOrUpdateKubeBase
         [PSArgumentCompleter("Delete", "Deallocate")]
         public string NodeScaleSetEvictionPolicy { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "Add-on names to be enabled when cluster is created.")]
+        [ValidateNotNullOrEmpty()]
+        [PSArgumentCompleter("HttpApplicationRouting", "Monitoring", "VirtualNode", "AzurePolicy", "KubeDashboard")]
+        public string[] EnableAddOnName { get; set; }
```
Comment (erich-wang) [inline src/Aks/Aks/Commands/NewKubeBase.cs:69]: better naming?

File: src/Aks/Aks/Commands/UpdateAddonsBase.cs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Id of a managed Kubernetes cluster.")]
+        [ValidateNotNullOrEmpty]
+        [Alias("ResourceId")]
+        public string Id { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Add-on names to be enabled when cluster is created.")]
+        [ValidateNotNullOrEmpty()]
+        [PSArgumentCompleter("HttpApplicationRouting", "Monitoring", "VirtualNode", "AzurePolicy", "KubeDashboard")]
+        public string[] AddOnName { get; set; }
```
Comment (erich-wang) [inline src/Aks/Aks/Commands/UpdateAddonsBase.cs:72]: Suggest to use `Name`

**Review by erich-wang — CHANGES_REQUESTED** [review]

**Review by erich-wang — DISMISSED** [review]

**Comment by dingmeng-xue** [comment]
Merge it and only windows test is failed due to time out.

**Pull request merged**